### PR TITLE
21 allow to run app from jar

### LINKED ISSUE #21: The app does not run with java -jar
## Description

When trying to run the app using an executable jar I get ClassNotFoundException.

## How to reproduce

```
$ mvn clean package
[...]
$ java -jar target/tetris-0.1.jar 
Error: Could not find or load main class com.epam.prejap.git.tetris.Tetris
Caused by: java.lang.ClassNotFoundException: com.epam.prejap.git.tetris.Tetris
```

## Expected behavior

```
$ mvn clean package
[...]
$ java -jar target/tetris-0.1.jar
+--------------------+
|         ##         |
|         ##         |
|                    |
|                    |
|                    |
|                    |
|                    |
|                    |
|                    |
|                    |
+--------------------+
[...]

### PR DESCRIPTION
**Pull request by konopkagrzegorz**
We have finished issue with not running app from jar.

We have changed:
- pom.xml - bumped version and added maven-surefile-plugin
- created ExecutableJarIT class with appropriate tests

To run test type in terminal in project's directory where pom.xml is located:
```mvn verify```

Closes #21

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -38,11 +39,38 @@
...
+                <executions>
+                    <execution>
+                        <id>integration-test</id>
+                        <goals>
+                            <goal>test</goal>
+                        </goals>
+                        <phase>integration-test</phase>
+                        <configuration>
+                            <excludes>
+                                <exclude>none</exclude>
```
Comment (pkierat-epam) [inline pom.xml:65]: This is not very useful. Maven is built around "Convention over configuration" principle.

File: pom.xml [hunk 2]
```diff
@@ -38,11 +39,38 @@
...
                             <addClasspath>true</addClasspath>
-                            <mainClass>com.epam.prejap.tetris.Tetris</mainClass>
+                            <mainClass>com.epam.prejap.teatrees.TeaTrees</mainClass>
                         </manifest>
                     </archive>
                 </configuration>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
```
Comment (pkierat-epam) [inline pom.xml:49]: Maven Surefire plugin is used to run unit tests. For IT, the plugin is called Failsafe.

File: src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import static org.testng.Assert.*;
+
+@Test
+public class ExecutableJarIT {
+
+    private static final String MAIN_CLASS_FQN = TeaTrees.class.getName();
+    private static String CLASSPATH;
+    private static final String JAR_EXTENSION = "jar";
+
+    static {
```
Comment (pkierat-epam) [inline src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java:22]: Why static initialization block?

File: src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        CodeSource codeSource = TeaTrees.class.getProtectionDomain().getCodeSource();
+        File jarFile;
+        try {
+            jarFile = new File(codeSource.getLocation().toURI().getPath());
+            CLASSPATH = jarFile.getParentFile().getPath();
+        } catch (URISyntaxException e) {
+            e.getMessage();
```
Comment (pkierat-epam) [inline src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java:29]: This line does absolutely nothing. (It invokes a method and ignores its result.)

File: src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        for (Path path : paths) {
+            JarFile jarFile = new JarFile(String.valueOf(path));
+            assertNotNull(jarFile.getManifest(), "MANIFEST.MF does not exist");
+        }
+    }
+
+    @Test(dependsOnMethods = "testJARisCreated")
+    public void manifestFileContainsMainClass() throws IOException {
+        String expected = MAIN_CLASS_FQN;
+        List<Path> paths = Files.
```
Comment (pkierat-epam) [inline src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java:57]: This code is duplicated. Please remove the redudancy.

**Review by pkierat-epam — CHANGES_REQUESTED** [review]

**Comment by kidolega** [comment]
<plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-jar-plugin</artifactId>
                <version>3.2.0</version>
                <configuration>
                    <archive>
                        <manifest>
                            <addClasspath>true</addClasspath>
                            <mainClass>com.epam.prejap.teatrees.TeaTrees</mainClass>
                        </manifest>
                    </archive>
                </configuration>
            </plugin>

wystarczy taki plugin zeby odpalic to z jara

File: src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.CodeSource;
+import java.util.List;
+import java.util.jar.JarFile;
+
+import static org.testng.Assert.*;
+
+@Test
+public class ExecutableJarIT {
```
Comment (Nimrud) [inline src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java:17]: No javadoc

Comment (konopkagrzegorz) [reply src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java:17]: Javadoc for Test class?

File: src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        List<Path> paths = findPathsWithExtensions(CLASSPATH,JAR_EXTENSION);
+
+        for (Path path : paths) {
+            JarFile jarFile = new JarFile(String.valueOf(path));
+            String actual = jarFile.getManifest().getMainAttributes().getValue("Main-Class");
+            assertEquals(actual, expected);
+        }
+    }
+
+    private List<Path> findPathsWithExtensions(String path, String extension) throws IOException {
```
Comment (emmaline261b) [inline src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java:63]: Nice extraction!

File: src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java (hunk 6 above)
Comment (pkierat-epam) [reply src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java:17]: I see no point in documenting tests, unless there's something non-obvious which needs to be explicitly noted.

Comment (pkierat-epam) [reply src/test/java/com/epam/prejap/teatrees/ExecutableJarIT.java:17]: I see no point in documenting tests, unless there's something non-obvious which needs to be explicitly noted.

**Pull request merged**